### PR TITLE
BUG: Fix #4018 `reconstruct_path` for trivial graph

### LINKED ISSUE #4018: reconstruct_path for degenerate graph
`scipy.sparse.csgraph.breadth_first_tree` raises an exception when the input is a degenerate graph that consists of a single node with no edge (or one with a self loop; it does not matter). Is it as intended? I would expect for it to return the input graph itself...

I don't know if it is a "bug", but let me just report what I found.

1.
The error I got:
```
>>> from scipy.sparse import csr_matrix
>>> from scipy.sparse.csgraph import breadth_first_tree
>>> A = csr_matrix((1, 1))  # Degenerate graph
>>> A.todense()
matrix([[ 0.]])
>>> breadth_first_tree(A, i_start=0)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "_traversal.pyx", line 162, in scipy.sparse.csgraph._traversal.breadth_first_tree (scipy/sparse/csgraph/_traversal.c:2159)
  File "_tools.pyx", line 352, in scipy.sparse.csgraph._tools.reconstruct_path (scipy/sparse/csgraph/_tools.c:4229)
  File "/usr/local/lib/python2.7/site-packages/scipy/sparse/compressed.py", line 88, in __init__
    self.check_format(full_check=False)
  File "/usr/local/lib/python2.7/site-packages/scipy/sparse/compressed.py", line 167, in check_format
    raise ValueError("indices and data should have the same size")
ValueError: indices and data should have the same size
```
Here I would expect to get a sparse matrix that is equal to `A`.

`breadth_first_tree` is actually `breadth_first_order` + `reconstruct_path`:
```
>>> from scipy.sparse.csgraph import breadth_first_order, reconstruct_path
>>> node_array, predecessors = breadth_first_order(A, i_start=0)
>>> node_array
array([0], dtype=int32)
>>> predecessors
array([-9999], dtype=int32)
>>> reconstruct_path(A, predecessors)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "_tools.pyx", line 352, in scipy.sparse.csgraph._tools.reconstruct_path (scipy/sparse/csgraph/_tools.c:4229)
  File "/usr/local/lib/python2.7/site-packages/scipy/sparse/compressed.py", line 88, in __init__
    self.check_format(full_check=False)
  File "/usr/local/lib/python2.7/site-packages/scipy/sparse/compressed.py", line 167, in check_format
    raise ValueError("indices and data should have the same size")
ValueError: indices and data should have the same size
```

2.
Look into [the code](https://github.com/scipy/scipy/blob/master/scipy/sparse/csgraph/_tools.pyx#L298):
```
>>> from scipy.sparse.csgraph._validation import validate_graph
>>> import numpy as np
>>> ITYPE = np.int32
>>> csgraph = validate_graph(A, directed=True, dense_output=False)
>>> N = csgraph.shape[0]
>>> nnull = (predecessors < 0).sum()
>>> indices = np.argsort(predecessors)[nnull:].astype(ITYPE)
>>> pind = predecessors[indices]
>>> indptr = pind.searchsorted(np.arange(N + 1)).astype(ITYPE)
>>> data = csgraph[pind, indices]
```

Variables defined:
```
>>> csgraph
<1x1 sparse matrix of type '<type 'numpy.float64'>'
	with 0 stored elements in Compressed Sparse Row format>
>>> indices
array([], dtype=int32)
>>> pind
array([], dtype=int32)
>>> indptr
array([0, 0], dtype=int32)
>>> data
<1x0 sparse matrix of type '<type 'numpy.float64'>'
	with 0 stored elements in Compressed Sparse Row format>
```
Note that `data` is a 1x0 sparse matrix.
Next line:
```
>>> data = np.asarray(data).ravel()
>>> data
array([ <1x0 sparse matrix of type '<type 'numpy.float64'>'
	with 0 stored elements in Compressed Sparse Row format>], dtype=object)
```
Here's where the error arises:
```
>>> csr_matrix((data, indices, indptr), shape=(N, N))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/site-packages/scipy/sparse/compressed.py", line 88, in __init__
    self.check_format(full_check=False)
  File "/usr/local/lib/python2.7/site-packages/scipy/sparse/compressed.py", line 167, in check_format
    raise ValueError("indices and data should have the same size")
ValueError: indices and data should have the same size
```
If `data` was an empty numpy.matrix, there would be no error:
```
>>> empty_matrix = np.matrix([[]])
>>> empty_matrix = np.asarray(empty_matrix).ravel()
>>> B = csr_matrix((empty_matrix, indices, indptr), shape=(N, N))
>>> B
<1x1 sparse matrix of type '<type 'numpy.float64'>'
	with 0 stored elements in Compressed Sparse Row format>
>>> B.todense()
matrix([[ 0.]])
```

3.
Sparse matrix exhibits somehow "inconsistent" behavior.
Consider for example the `A` matrix again:
```
>>> A
<1x1 sparse matrix of type '<type 'numpy.float64'>'
	with 0 stored elements in Compressed Sparse Row format>
>>> A.todense()
matrix([[ 0.]])
```
Indexing of `A` with nonempty arrays returns a numpy matrix:
```
>>> A[[0], [0]]
matrix([[ 0.]])
```
But indexing with empty arrays returns a sparse matrix:
```
>>> A[[], []]
<1x0 sparse matrix of type '<type 'numpy.float64'>'
	with 0 stored elements in Compressed Sparse Row format>
>>> A[[], []].todense()
matrix([], shape=(1, 0), dtype=float64)
```
If `A[[], []]` too is an (empty) numpy.matrix, then the above error would not have occurred.

4.
The version of scipy:
```
>>> import scipy
>>> scipy.__version__
'0.14.0'
```

### PR DESCRIPTION
**Pull request by oyamad**
Fix #4018, and add tests with a trivial graph.
- Use `np.matrix.getA1` instead of `np.asarray().ravel()`; it raises an exception if `data` is a sparse matrix, which it is the case when the input graph is a trivial graph (a graph with a single node).
- Add tests for `breadth_first_tree` and `depth_first_tree` for the case where the input graph is a trivial graph.

### CODE REVIEW

File: scipy/sparse/csgraph/_tools.pyx [hunk 1]
```diff
@@ -334,16 +334,25 @@ def reconstruct_path(csgraph, predecessors, directed=True):
...
-    else:
-        data1 = csgraph[pind, indices]
+    data = csgraph[pind, indices]
+
+    # Fix issue #4018:
+    # If `pind` and `indices` are empty arrays, `data` is a sparse matrix
+    # (it is a numpy.matrix otherwise); handle this case separately.
+    try:
+        data = data.getA1()
+    except:
```
Comment (larsmans) [inline scipy/sparse/csgraph/_tools.pyx]: Please don't use a bare `except`. Any exception other than `AttributeError` could signal a bug.

In fact, I would use an explicit type check here:

```py
if isspmatrix(data):
    data = data.todense()
```

File: scipy/sparse/csgraph/_tools.pyx [hunk 2]
```diff
@@ -334,16 +334,25 @@ def reconstruct_path(csgraph, predecessors, directed=True):
...
+        data = data.todense().getA1()
+
+    if directed is False:
```
Comment (larsmans) [inline scipy/sparse/csgraph/_tools.pyx]: `if not directed`

**Comment by oyamad** [comment]
@larsmans Thank you for your comments.

I corrected the code according to them.

**Comment by larsmans** [comment]
Looks good now, merging.

**Pull request merged**